Add permissions guide (#278)
mirumee/saleor-docs: docs/developer/permissions.mdx

---
title: Permissions
sidebar_label: Permissions
---

## User permissions

> Its group defines user permissions. Organizing access rights into [Grups](developer/api-reference/objects/group.mdx) helps in determining staff users' roles in the organization.

Reviewer: ```suggestion
Its group defines user permissions. Organizing access rights into [Groups](developer/api-reference/objects/group.mdx) helps in determining staff users' roles in the organization.
```

> Its group defines user permissions. Organizing access rights into [Groups](developer/api-reference/objects/group.mdx) helps in determining staff users' roles in the organization.

Reviewer 2: "Its group defines user permissions." - What do you mean exactly? I'm not sure I understand this sentence.

Author: You cannot assign permissions to the user directly, only by assigning user to permission group

Reviewer 2: Ah so this makes it more clear, maybe the explanation you gave should be included in the docs instead?

Author: ```suggestion
Instead of assigning permissions directly to the user, we define them on a by-group basis. Organizing access rights into [Groups](developer/api-reference/objects/group.mdx) helps in determining staff users' roles in the organization.
```

> Instead of assigning permissions directly to the user, we define them on a by-group basis. Organizing access rights into [Groups](developer/api-reference/objects/group.mdx) helps in determining staff users' roles in the organization.

Reviewer 3: ```suggestion
Instead of assigning permissions directly to the user, we define them on a group basis. Organizing access rights in [Groups](developer/api-reference/objects/group.mdx) helps in defining the roles of team members.
```

Examples of groups:
- Translators - `MANAGE_TRANSLATIONS` permission.
- Customer support - `MANAGE_ORDERS` and `MANAGE_USERS` permissions.

### Creating and removing groups
To create a new group, use [permissionGroupCreate](developer/api-reference/mutations/permission-group-create.mdx).

Request:
```graphql
mutation {
  permissionGroupCreate(
    input: {
      addPermissions: [MANAGE_GIFT_CARD, MANAGE_DISCOUNTS]

>       addUsers: []

Reviewer 2: Since we include this field in the example, we could also describe its behavior somewhere, for example: 
"When creating a new permission group you can optionally assign a list of members using the `addUsers` input field."

Author: addUsers field is explained in the managing members section, IMO there's no need to explain it twice

      name: "Sale managers"
    }
  ) {
    errors {
      message
    }
    group {
      id
      name
      permissions {
        name
      }
    }
  }
}
```

Successful response:
```json
{
  "data": {
    "permissionGroupCreate": {
      "errors": [],
      "group": {
        "id": "R3JvdXA6NDM=",
        "name": "Sale managers",
        "permissions": [
          {
            "name": "Manage sales and vouchers.",
          },
          {
            "name": "Manage gift cards.",
          }
        ]
      },
    }
  }
}
```

> To remove group use [permissionGroupDelete](developer/api-reference/mutations/permission-group-delete.mdx) mutation.

Reviewer 3: To remove group use [permissionGroupDelete](developer/api-reference/mutations/permission-group-delete.mdx) mutation.

> To remove group, use [permissionGroupDelete](developer/api-reference/mutations/permission-group-delete.mdx) mutation.

Reviewer 4: IMO would be nice to have similar example for removing as we have for creating group

Author: Felt that adding an example for that doesn't add anything to the document besides making it longer.

### Modifying group and managing its members

> [permissionGroupUpdate](developer/api-reference/mutations/permission-group-update.mdx) takes a list of user IDs you would like to add or remove from the group. Having a user in both of the lists will result in an error.

Reviewer: ```suggestion
[permissionGroupUpdate](developer/api-reference/mutations/permission-group-update.mdx) takes a list of user IDs you would like to add or remove from the group. Having a same user in both of the lists will result in an error.
```

Example request:
```graphql
mutation {
  permissionGroupUpdate(
    id: "R3JvdXA6NDM="
    input: {
      name: "Sale managers"
      addPermissions: []
      removePermissions: []
      addUsers: ["VXNlcjozMg=="]
      removeUsers: []
    }
  ) {
    errors {
      message
    }
  }
}
```

## App permissions

In contrast to users, app permissions are defined per-app basis. Access can be assigned during app installation and modified later using [appUpdate](developer/api-reference/mutations/app-update.mdx) mutation.

## JWT token and permissions

> JWT tokens have a list of assigned permissions. Decoding its contents using HS256 algorithm you will get:

Reviewer: HS256 is a signature algorithm, did you mean to decode the JWT token payload instead?

Author: Thats what I meant. Good feedback 👍

Author: ```suggestion
JWT tokens have a list of assigned permissions. By decoding payload using HS256 algorithm you will get:
```

```json
{
  "iat": 1624013260,
  "iss": "example.com",
  "token": "AixxXXXxzF",
  "email": "john@example.com",
  "type": "access",
  "user_id": "VXNlcjozMg==",
  "is_staff": true,
  "exp": 1624049260,
  "oauth_access_key": "",
  "permissions": [
    "MANAGE_TRANSLATIONS",
    "MANAGE_PRODUCTS",
    "MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES"
  ]
}
```

To check token online and learn more about JWT visit [https://jwt.io](https://jwt.io).

## Available permissions

> Available permissions are kept in the [PermissionEnum](developer/api-reference/enums/permission-enum.mdx).

Author: Why add enum, if its in the name?

Examples of groups:
- Translators - `MANAGE_TRANSLATIONS` permission.
- Customer support - `MANAGE_ORDERS` and `MANAGE_USERS` permissions.

### Creating and removing groups
To create a new group, use [permissionGroupCreate](developer/api-reference/mutations/permission-group-create.mdx).

Request:
```graphql
mutation {
  permissionGroupCreate(
    input: {
      addPermissions: [MANAGE_GIFT_CARD, MANAGE_DISCOUNTS]
      addUsers: []
      name: "Sale managers"
    }
  ) {
    errors {
      message
    }
    group {
      id
      name
      permissions {
        name
      }
    }
  }
}
```

Successful response:
```json
{
  "data": {
    "permissionGroupCreate": {
      "errors": [],
      "group": {
        "id": "R3JvdXA6NDM=",
        "name": "Sale managers",
        "permissions": [
          {
            "name": "Manage sales and vouchers.",
          },
          {
            "name": "Manage gift cards.",
          }
        ]
      },
    }
  }
}
```

> To remove group, use [permissionGroupDelete](developer/api-reference/mutations/permission-group-delete.mdx) mutation:

Reviewer 3: ```suggestion
To remove a group, use the [permissionGroupDelete](developer/api-reference/mutations/permission-group-delete.mdx) mutation:
```

Example request:
```graphql
mutation {
  permissionGroupUpdate(
    id: "R3JvdXA6NDM="
    input: {
      name: "Sale managers"
      addPermissions: []
      removePermissions: []
      addUsers: ["VXNlcjozMg=="]
      removeUsers: []
    }
  ) {
    errors {
      message
    }
  }
}
```

## App permissions

In contrast to users, app permissions are defined per-app basis. Access can be assigned during app installation and modified later using [appUpdate](developer/api-reference/mutations/app-update.mdx) mutation.

## JWT token and permissions

JWT tokens have a list of assigned permissions. By decoding payload using HS256 algorithm you will get:

```json
{
  "iat": 1624013260,
  "iss": "example.com",
  "token": "AixxXXXxzF",
  "email": "john@example.com",
  "type": "access",
  "user_id": "VXNlcjozMg==",
  "is_staff": true,
  "exp": 1624049260,
  "oauth_access_key": "",
  "permissions": [
    "MANAGE_TRANSLATIONS",
    "MANAGE_PRODUCTS",
    "MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES"
  ]
}
```

To check token online and learn more about JWT visit [https://jwt.io](https://jwt.io).

:::note

> Since Saleor reads permissions from the JWT, generating new token is nessesary when user permissions change.

Reviewer 3: ```suggestion
Since Saleor reads permissions from the JWT token, generating a new token is necessary when the user changes permissions.
```

Author: JWT stands for Json Web Token. No need to add token gain
